Core/Src/bmb.c

```diff
@@ -236,6 +287,84 @@ void updateBmbData(Bmb_S* bmb, uint32_t numBmbs)
 			bmb[j].blockV = blockV;
 		}
 	}
+	else
+	{
+		printf("Error during VBLOCK readAll!\r\n");
+
+		// Failed to acquire data. Set status to MIA
+		for (int j = 0; j < numBmbs; j++)
+		{
+			bmb[j].blockVStatus = MIA;
+		}
+	}
+
+	// TODO Add check - Compare VBLOCK with sum of brick voltages
+}
+
+void updateBMBTempData(Bmb_S* bmb, uint32_t numBmbs)
+{
+	// Cycle through MUX channels
+	setMux(numBmbs, MUX1);
+	for(int i = MUX1; i < NUM_MUX_CHANNELS; i++)
+	{
+		// Read AUX/TEMP registers
+		for (int auxChannel = AIN1; auxChannel <= AIN2; auxChannel++)
+		{
+			if (readAll(auxChannel, recvBuffer, numBmbs))
+			{
+				// Parse received data
+				for (uint8_t j = 0; j < numBmbs; j++)
+				{
+					// Read AUX voltage in [15:4]
+					uint32_t auxRaw = ((recvBuffer[4 + 2*j] << 8) | recvBuffer[3 + 2*j]) >> 4;
+					float auxV = auxRaw * CONVERT_12BIT_TO_3V3;
+					bmb[j].tempVoltage[muxState + ((auxChannel == AIN2) ? NUM_MUX_CHANNELS : 0)] = auxV;
+
+					// Convert temp voltage registers to temperature readings
+					if(muxState == (MUX7 || MUX8)) // NTC/ON-Board Temp Channel
+					{
+						bmb[j].boardTemp[muxState - MUX7 + ((auxChannel == AIN2) ? 2 : 0)] = lookup(auxV, &ntcTable);
+					}
+					else // Zener/Brick Temp Channel
+					{
+						bmb[j].brickTemp[muxState + ((auxChannel == AIN2) ? NUM_MUX_CHANNELS : 0)] = lookup(auxV, &zenerTable);
+					}
+				}
+			}
+			else
+			{
+				// Failed to acquire data. Set status to MIA
+				for (int j = 0; j < numBmbs; j++)
+				{
+					bmb[j].tempStatus[muxState + ((auxChannel == AIN2) ? 8 : 0)] = MIA;
+				}
+			}
+		}
+	}
+}
+
+void setMux(uint32_t numBmbs, uint8_t muxSetting)
+{
+	bool gpio[3];
+	for(int i = 0; i < 3; i++)
+	{
+		gpio[i] = (muxSetting >> i) & 1;
+	}
+	setGpio(numBmbs, gpio[0], gpio[1], gpio[2], gpio3State); // Currently sets GPIO 4 to 0 when updating MUX
+}
+
+void setGpio(uint32_t numBmbs, bool gpio0, bool gpio1, bool gpio2, bool gpio3)
+{
+	// First 4 bits set GPIO to output mode
+	// Last 4 bits set GPIO logic state for channels 3, 2, 1, 0 respectively
+	uint16_t data = 0xF000 | (gpio0) | (gpio1 << 1) | (gpio2 << 2) | (gpio3 << 3);
```

Comment:
Personally I would think it would be clearer if you had it in order of the bits. gpio3 gpio2 gpio1 and gpio0 in that order from left to right